src/devices/bresser_leakage.c

```diff
@@ -0,0 +1,169 @@
+ [Reset]
+ C0 10 36 79 37 09 51 70 00 00 00 00 00 00 00 00 01 1E FD FD FF FF FF DF FF FF [CH1]
+ C0 10 36 79 37 09 51 B0 00 00 00 00 00 00 00 00 03 FE FD FF AF FF FF FF FF FD [CH1+ALARM]
+ [Reset]
+ 71 9C 54 81 72 09 51 40 00 00 00 00 00 00 00 00 0F FF FF FF FF FF FF DF FF FE [CH1+BATT_LO]
+ 71 9C 54 81 72 09 51 40 00 00 00 00 00 00 00 00 0F FE FF FF FF FF FB FF FF FF
+ 71 9C 54 81 72 09 51 40 00 00 00 00 00 00 00 00 07 FD F7 FF DF FF FF DF FF FF
+ 71 9C 54 81 72 09 51 80 00 00 00 00 00 00 00 00 1F FF FF F7 FF FF FF FF FF FF [CH1+BATT_LO+ALARM]
+ F0 94 54 81 72 09 59 40 00 00 00 00 00 00 00 00 0F FF DF FF FF FF FF BF FD F7 [CH1+BATT_LO+NSTARTUP]
+ F0 94 54 81 72 09 59 80 00 00 00 00 00 00 00 00 03 FF B7 FF ED FF FF FF DF FF [CH1+BATT_LO+NSTARTUP+ALARM]
+
+ */
+
+#include "decoder.h"
+
+static int bresser_leakage_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t const preamble_pattern[] = {0xaa, 0xaa, 0x2d, 0xd4};
+    data_t *data;
+    uint8_t msg[18];
+    uint32_t sensor_id;
+    uint8_t s_type;
+    uint8_t chan;
+    uint8_t startup;
+    uint8_t battery_ok;
+    int alarm;
+    int no_alarm;
+    int decode_ok;
+    uint8_t null_bytes;
+
+    if (bitbuffer->num_rows != 1
+            || bitbuffer->bits_per_row[0] < 160
+            || bitbuffer->bits_per_row[0] > 440) {
+        decoder_logf(decoder, 2, __func__, "bit_per_row %u out of range", bitbuffer->bits_per_row[0]);
+        return DECODE_ABORT_EARLY; // Unrecognized data
+    }
+
+    unsigned start_pos = bitbuffer_search(bitbuffer, 0, 0,
+            preamble_pattern, sizeof (preamble_pattern) * 8);
+
+    if (start_pos >= bitbuffer->bits_per_row[0]) {
+        return DECODE_ABORT_LENGTH;
+    }
+    start_pos += sizeof (preamble_pattern) * 8;
+
+    unsigned len = bitbuffer->bits_per_row[0] - start_pos;
+    if (len < sizeof(msg) * 8) {
+        decoder_logf(decoder, 2, __func__, "%u too short", len);
+        return DECODE_ABORT_LENGTH; // message too short
+    }
+
+    bitbuffer_extract_bytes(bitbuffer, 0, start_pos, msg, sizeof(msg) * 8);
+
+    decoder_log_bitrow(decoder, 2, __func__, msg, sizeof(msg) * 8, "");
+
+    // TODO: Find parity/checksum/checksum algorithm
+
+    sensor_id     = ((uint32_t)msg[2] << 24) | (msg[3] << 16) | (msg[4] << 8) | (msg[5]);
+    s_type        = msg[6] >> 4;
+    chan          = (msg[6] & 0x7);
+    battery_ok    = ((msg[7] & 0x30) != 0x00) ? 1 : 0;
+    startup       = (msg[6] >> 3) & 1;
+    alarm         = ((msg[7] & 0x80) == 0x80) ? 1 : 0;
+    no_alarm      = ((msg[7] & 0x40) == 0x40) ? 1 : 0;
+
+
+    null_bytes = msg[7] & 0xF;
+
+    for (int i=8; i<=15; i++) {
+        null_bytes |= msg[i];
+    }
+
+    // The parity/checksum/digest algorithm is currently unknown.
+    // We apply some heuristics to validate that the message is really from
+    // a water leakage sensor.
+    decode_ok = (s_type == SENSOR_TYPE_LEAKAGE) &&
+                (alarm != no_alarm) &&
```

Comment:
Continued line indent is 8 spaces, not aligned.

Reply:
Fixed